runtime/bcutil/defineclass.c

```diff
@@ -186,11 +189,19 @@ internalDefineClass(
 						romClassHashTableAdd(hashTable, romClass);
 					}
 				}
+				omrthread_monitor_exit(vm->classTableMutex);
 			}
 		} else if (NULL != orphanROMClass) {
 			/* ramClass creation succeeded - the orphanROMClass is no longer an orphan */
 			Trc_BCU_romClassOrphansHashTableDelete(vmThread, classNameLength, className, orphanROMClass);
-			romClassHashTableDelete(classLoader->romClassOrphansHashTable, orphanROMClass);
+			/* All access to the orphan table must be done while holding classTableMutex.
+			 * Ensure the entry is still in the table before removing it.
+			 */
+			omrthread_monitor_enter(vm->classTableMutex);
+			if (romClassHashTableFind(classLoader->romClassOrphansHashTable, className, classNameLength) == orphanROMClass) {
```

Comment:
Should there be a tracepoint for the case where the orphan wasn't in the orphan's table?  It would give us something to trigger on if this unlikely (impossible) situation occurred.... though writing that makes me think assert rather than tracepoint

Reply:
Tracepoint added.  Assert makes me a little leary as it's possible to disable the parallel class loading table via a command-line option.